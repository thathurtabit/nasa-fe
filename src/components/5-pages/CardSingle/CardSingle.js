import React, { Component, Suspense, lazy } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import CardSingleStyled from './CardSingle.styled';
import Loading from '../../1-atoms/Loading/Loading';
import { NoTitle, NoDesc } from '../../../utils/constants/constants';

const NoItems = lazy(() => import('../../2-molecules/NoItems/NoItems'));
const Card = lazy(() => import('../../3-organisms/Card/Card'));

const mapStateToProps = state => ({
  api: state.api,
  products: state.response,
});

export class CardSingle extends Component {
  constructor(props) {
    super(props);
    this.state = { isLoading: true, error: false };
  }

  componentDidMount() {
    this.mounted = true;
    this.handleFetchData = this.handleFetchData.bind(this);
    this.handleFetchError = this.handleFetchError.bind(this);

    const { api, products } = this.props;

    if (!products.length && this.mounted) {
      fetch(api, { credentials: 'same-origin', method: 'GET' })
        .then(response => response.json())
        .then(data => this.handleFetchData(data))
        .catch(error => this.handleFetchError(error));
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  handleFetchData(data) {
    const { location, products } = this.props;
    const productId = location
      .split('/')
      .filter(loc => loc)
      .pop();

    if (this.mounted && !products.length) {
      // Filter out single Card
      const productSingle =
        data.Products.filter(prod => prod.MoonpigProductNo === productId)[0] ||
        null;
      // Keep what we need
      const product = {
        title: productSingle.Title || NoTitle,
        desc: productSingle.ShortDescription || NoDesc,
        imgSrc: productSingle.ProductImage.Link.Href,
        id: productSingle.ProductId,
        productNo: productSingle.MoonpigProductNo,
        link: '/',
      };

      this.setState({ product, isLoading: false });
    }
  }

  handleFetchError(error) {
    this.setState({ error: true, isLoading: false });
    console.warn(error);
  }

  render() {
    const { isLoading, product, error } = this.state;

    return (
      <CardSingleStyled>
        {isLoading && <Loading isLoading />}
        <Suspense fallback={<Loading isLoading={isLoading} />}>
          {error ? (
            <NoItems text="No items found." />
          ) : (
            !isLoading && product !== null && <Card product={product} />
          )}
        </Suspense>
      </CardSingleStyled>
    );
  }
}

export default connect(mapStateToProps)(CardSingle);

CardSingle.propTypes = {
  location: PropTypes.string.isRequired,
  api: PropTypes.string.isRequired,
  products: PropTypes.arrayOf(
    PropTypes.objectOf(
      PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
        PropTypes.object,
      ])
    )
  ),
};

CardSingle.defaultProps = {
  products: [],
};
