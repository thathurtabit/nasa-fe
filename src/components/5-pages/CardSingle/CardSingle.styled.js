import styled from 'styled-components';

const CardSingleStyled = styled.section`
  background: ${props => props.theme.colors.header};
  margin: 0 auto;
  max-width: ${props => props.theme.breakpoints.sm};
  position: relative;

  @media screen and (min-width: ${props => props.theme.breakpoints.md}) {
    max-width: ${props => props.theme.breakpoints.md};
  }
`;

export default CardSingleStyled;
