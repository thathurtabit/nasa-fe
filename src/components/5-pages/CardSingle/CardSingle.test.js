import React from 'react';
import { shallowToJson } from 'enzyme-to-json';
import 'jest-styled-components';
import { shallowWithTheme } from '../../../utils/test_config/testHelpers';

import { CardSingle } from './CardSingle';

describe('CardSingle', () => {
  it('should render correctly', () => {
    const element = shallowWithTheme(<CardSingle location="url" api="api" />);
    expect(shallowToJson(element)).toMatchSnapshot();
  });
});
