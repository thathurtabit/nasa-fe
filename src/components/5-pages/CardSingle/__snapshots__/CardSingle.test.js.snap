// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardSingle should render correctly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
